File: protect/src/main/java/com/webileapps/safeguard/SignatureComparison.java
package com.webileapps.safeguard;

import android.content.Context;
import android.content.pm.PackageManager;
import android.content.pm.Signature;
import android.content.pm.SigningInfo;
import android.os.Build;
import android.util.Log;
import java.security.MessageDigest;
import java.security.NoSuchAlgorithmException;

public class SignatureComparison {
    private static final String TAG = "AppSignatureVerifier";
    private static final String HASH_ALGORITHM = String.join("-", "SHA", "256");
    
    public boolean isAppSignatureValid(Context context, String expectedSignatureHash) {
        try {
            // Validate input parameters
            if (context == null || expectedSignatureHash == null) {
                throw new IllegalArgumentException("Null context or expected hash");
            }

            final String packageName = context.getPackageName();
            final Signature signature = getAppSignature(context, packageName);
            
            if (signature == null) {
                Log.e(TAG, "No valid signature found");
                return false;
            }

            final String currentHash = calculateSignatureHash(signature);
            final String normalizedExpected = normalizeHash(expectedSignatureHash);
            final String normalizedCurrent = normalizeHash(currentHash);

            final boolean isValid = constantTimeEquals(normalizedExpected, normalizedCurrent);
            
            if (!isValid) {
                Log.w(TAG, "Signature mismatch!\n" +
                        "Expected: " + normalizedExpected + "\n" +
                        "Actual:   " + normalizedCurrent);
            }
            
            return isValid;
        } catch (Exception e) {
            Log.e(TAG, "Signature verification failed: " + e.getMessage());
            return false;
        }
    }

    private Signature getAppSignature(Context context, String packageName) 
        throws PackageManager.NameNotFoundException {
        final PackageManager pm = context.getPackageManager();
        final int flags = Build.VERSION.SDK_INT >= Build.VERSION_CODES.P ?
            PackageManager.GET_SIGNING_CERTIFICATES :
            PackageManager.GET_SIGNATURES;

        final android.content.pm.PackageInfo packageInfo = pm.getPackageInfo(packageName, flags);
        
        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
            final SigningInfo signingInfo = packageInfo.signingInfo;
            return (signingInfo != null && signingInfo.hasMultipleSigners()) ? 
                signingInfo.getApkContentsSigners()[0] : 
                signingInfo != null ? signingInfo.getSigningCertificateHistory()[0] : null;
        }
        return packageInfo.signatures != null ? packageInfo.signatures[0] : null;
    }

    private String calculateSignatureHash(Signature signature) 
        throws NoSuchAlgorithmException {
        final MessageDigest md = MessageDigest.getInstance(HASH_ALGORITHM);
        final byte[] hashBytes = md.digest(signature.toByteArray());
        return bytesToHex(hashBytes);
    }

    // Constant-time comparison to prevent timing attacks
    private boolean constantTimeEquals(String a, String b) {
        if (a.length() != b.length()) {
            return false;
        }

        int result = 0;
        for (int i = 0; i < a.length(); i++) {
            result |= a.charAt(i) ^ b.charAt(i);
        }
        return result == 0;
    }

    private String normalizeHash(String hash) {
        return hash.replaceAll("[^A-Fa-f0-9]", "").toUpperCase();
    }

    private static String bytesToHex(byte[] bytes) {
        final char[] hexChars = new char[bytes.length * 2];
        for (int i = 0; i < bytes.length; i++) {
            final int v = bytes[i] & 0xFF;
            hexChars[i * 2] = "0123456789ABCDEF".charAt(v >>> 4);
            hexChars[i * 2 + 1] = "0123456789ABCDEF".charAt(v & 0x0F);
        }
        return new String(hexChars);
    }
}

File: protect/src/main/java/com/webileapps/safeguard/SignatureComparison.kt
package com.webileapps.safeguard

import android.content.Context
import android.content.pm.PackageManager
import android.os.Build
import android.util.Log
import java.security.MessageDigest


class SignatureComparison {

    val TAG: String = "AppSignatureVerifier"

    fun isAppSignatureValid(context: Context, expectedSignatureHash: String): Boolean {
        return try {
            val packageManager = context.packageManager
            val packageName = context.packageName
            val packageInfo = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                packageManager.getPackageInfo(packageName, PackageManager.GET_SIGNING_CERTIFICATES)
            } else {
                @Suppress("DEPRECATION")
                packageManager.getPackageInfo(packageName, PackageManager.GET_SIGNATURES)
            }

            val signatures = if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.P) {
                packageInfo.signingInfo?.apkContentsSigners
            } else {
                @Suppress("DEPRECATION")
                packageInfo.signatures
            }

            if (signatures == null || signatures.isEmpty()) {
                Log.e(TAG, "No signatures found for the app.")
                return false
            }


            val md = MessageDigest.getInstance("SHA-1");
            val originalDigest = md.digest(expectedSignatureHash.toByteArray())

            val currentDigest = md.digest(signatures[0].toByteArray())


            // Compare with the expected signature hash
            val isValid = MessageDigest.isEqual(originalDigest, currentDigest)
            if (isValid) {
                Log.d(TAG, "App signature is valid.")
            } else {
                Log.e(TAG, "App signature is invalid! Possible tampering detected.")
            }

            isValid
        } catch (e: Exception) {
            Log.e(TAG, "Error verifying app signature", e)
            false
        }
    }

    }